clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -972,16 +975,28 @@ private Record nextFetchedRecord() {
         private List<ConsumerRecord<K, V>> fetchRecords(int maxRecords) {
             if (isFetched)
                 return Collections.emptyList();
+            if (nextInlineException != null) {
```

Comment:
> On that I completely agree that we should just stick to that bad record and ask users to explicitly seek over it. But as you noticed, it is a behavioral change and probably worth discussing it in the community before we make that change (BTW, we do see some applications that just want to ignore the bad records if there are any. There are probably such use case in the community as well). So as for now, we may want to keep the behavior as is.

@becketqin The current behavior of the new consumer as of the last release of Kafka is to _not_ skip over messages which cause failures. My suggestion is the following:

1. We preserve that behavior for now. That means we must modify this patch to not skip over failed messages.
2. If you guys are convinced that messages with errors should be skipped, open a separate JIRA and let's have the discussion there on whether it's a good idea or not.

> Is this true? The exception is only associated with the corresponding bad record. In your example, if an exception is thrown from offset 1, that exception will be cleared. Fetching from offset 2 will only see an exception if offset 2 is a bad record, regardless of whether offset 1 was a bad record or not. So it seems not really depending on the previous exception.

My point is that the behavior of the consumer when its position is at 1 changes depending on what happened before. The current behavior, on the other hand, is deterministic: no matter what happened before, the next record that the consumer will attempt to return is the first one at an offset greater than or equal to 1. I am honestly not too sure how critical this is, but the inconsistent internal state makes me a bit uncomfortable (i.e. the fact that the record iterator is no longer aligned with the consumer's position and that we have no additional bookkeeping to know where it really is). Perhaps it's fine, but I'm reluctant to make the change until we've opened the issue publicly and given others a chance to weigh in. I have not fully thought about the compatibility implications for example.